app/plugin_requests.go

```diff
@@ -89,11 +89,24 @@ func (a *App) ServePluginPublicRequest(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	pluginID := vars["plugin_id"]
 
-	publicFilesPath, err := a.GetPluginsEnvironment().PublicFilesPath(pluginID)
+	pluginsEnv := a.GetPluginsEnvironment()
+
+	a.srv.PluginsLock.Lock()
+
+	// Check if someone has nullify the pluginsEnv in the meantime
+	if pluginsEnv == nil {
+		http.NotFound(w, r)
+		a.srv.PluginsLock.Unlock()
+		return
+	}
+
+	publicFilesPath, err := pluginsEnv.PublicFilesPath(pluginID)
 	if err != nil {
 		http.NotFound(w, r)
+		a.srv.PluginsLock.Unlock()
 		return
 	}
+	a.srv.PluginsLock.Unlock()
```

Comment:
Do we really need to lock? Once we've got the environment and verified it's not `nil` I am thinking we should be good.
Also if we go with the lock I believe it should be enough to read-lock.

Reply:
For the issues of an active server instance, we do not need the lock for sure. But there is a race condition during a shutdown. Good point regards the Rlock.

Reply:
What is the resource affected by the race? Getting the plugin environment through `GetPluginsEnvironment` is safe. If that returns a non-nil value then that's what we are eventually calling methods on. The server can still set to `nil` its own value but I don't see a big problem with that.

Reply:
Oh shoot, you are 💯 right. I thought the server value was the same pointer. Awesome, amending.